Sources/MongoSwiftSync/CursorCommon.swift

```diff
@@ -0,0 +1,42 @@
+/// A protocol describing the common behavior between cursor-like objects in the driver.
+internal protocol Cursor {
+    /// The decoded type iterated over by the cursor.
+    associatedtype T: Codable
+
+    /**
+     * Indicates whether this cursor has the potential to return more data.
+     *
+     * This property is mainly useful if this cursor is tailable, since in that case `tryNext` may return more results
+     * even after returning `nil`.
+     *
+     * For non-tailable cursors, the cursor will always be dead as soon as `tryNext` returns `nil` or an error.
+     */
+    var isAlive: Bool { get }
+
+    /**
+     * Get the next `T` from the cursor.
+     *
+     * If this cursor is tailable, this method will continue retrying until a non-empty batch is returned or the cursor
+     * is closed
+     */
+    func next() -> Result<T, Error>?
+
+    /**
+     * Attempt to get the next `T` from the cursor, returning `nil` if there are no results.
+     *
+     * If this cursor is tailable and `isAlive` is true, this may be called multiple times to attempt to retrieve more
+     * elements.
+     *
+     * If this cursor is a tailable await cursor, it will wait server side for a maximum of `maxAwaitTimeMS`
+     * before returning an empty batch. This option can be configured via options passed to the method that created this
+     * cursor (e.g. the `maxAwaitTimeMS` option on the `FindOptions` passed to `find`).
+     */
+    func tryNext() -> Result<T, Error>?
```

Comment:
We could make the signature of this throw and just return a `T?`, but I felt that it would be strange for `tryNext` to also have a different signature than `next` in addition to its different behavior. My worry was that a user might use `tryNext` thinking the signature difference was the reason for its existence rather than its behavioral differences. On the other hand, users shouldn't be invoking `next` manually in general, so the consistency isn't necessarily important in that respect. What do you think?

Reply:
We decided offline that `tryNext` should be okay due to precedents already existing for this type of thing (e.g. [`try` on `NSLock`](https://developer.apple.com/documentation/foundation/nslock/1418105-try))